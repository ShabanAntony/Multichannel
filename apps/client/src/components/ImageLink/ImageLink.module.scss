.base {
  & > img {
    object-fit: cover;
  }
  &__skeleton {
    background-color: rebeccapurple;
    opacity: 0.2;
  }
}
