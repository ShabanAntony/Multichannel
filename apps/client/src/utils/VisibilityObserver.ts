import { useEffect, useMemo, useRef } from 'react'

import { createEvent, createStore } from 'effector'
import { useUnit } from 'effector-react'

export const setVisibilityMap =
  createEvent<Map<Element, boolean>>('set visibility map')
export const observeElement = createEvent<Element>('starts element observing')
export const unobserveElement = createEvent<Element>('stop element observing')
export const $visibility = createStore(new Map<Element, boolean>()).on(
  setVisibilityMap,
  (_, store) => store,
)

const targetsMap = new WeakSet<Element>()
const observer = new IntersectionObserver(
  (entries) => {
    // eslint-disable-next-line effector/no-getState
    const oldMap = $visibility.getState()
    const map = entries.reduce((result, entry) => {
      result.set(entry.target, entry.isIntersecting)

      return result
    }, new Map<Element, boolean>(oldMap))

    setVisibilityMap(map)
  },
  {
    root: null,
    rootMargin: '0px',
    threshold: 0,
  },
)

observeElement.watch((element) => {
  // if (targetsMap.has(element)) {
  //   return
  // }
  // console.log(`>> add`, key)

  targetsMap.add(element)
  observer.observe(element)
})

unobserveElement.watch((element) => {
  targetsMap.delete(element)
  observer.unobserve(element)
})

export const useVisibilityObserver = () => {
  const containerRef = useRef(null)
  const visibility = useUnit($visibility)

  useEffect(() => {
    const element = containerRef.current!

    if (element) {
      observeElement(element)
    }

    return () => {
      if (element) {
        unobserveElement(element!)
      }
    }
  }, [])

  const isVisible = !!visibility.get(containerRef.current!)

  return useMemo(
    () => [containerRef, isVisible] as const,
    [containerRef, isVisible],
  )
}
