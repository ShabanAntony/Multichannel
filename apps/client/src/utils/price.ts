import {
  VSProfit,
  customsTax,
  nikeProfit,
  reebokProfit,
} from '../constants/prices'

export const formatUSDPrice = (price: number) => {
  return price + '$'
}

export const formatLaryPrice = (price: number) => {
  return Math.round(price) + ' áƒš'
}

export const calculateVSFinalPrice = (price: number) => {
  return Math.round(price * (1 + VSProfit + customsTax))
}

export const calculateNikeFinalPrice = (price: number) => {
  return Math.round(price * (1 + nikeProfit)) + 10
}

export const calculateReebokFinalPrice = (price: number) => {
  return Math.round(price * (1 + reebokProfit)) + 10
}
