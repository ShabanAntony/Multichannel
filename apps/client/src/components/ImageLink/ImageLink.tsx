import React, { FC, memo } from 'react'

import classNames from 'classnames'

import { useVisibilityObserver } from '../../utils/VisibilityObserver'

import styles from './ImageLink.module.scss'

export interface Props {
  readonly className?: string
  readonly link: string
  readonly src: string
  readonly alt: string
  readonly width: number
  readonly height: number
}

export const ImageLink: FC<Props> = memo(function ImageLink({
  className,
  link,
  src,
  alt,
  width,
  height,
}) {
  const [ref, isVisible] = useVisibilityObserver()
  const baseClasses = classNames(styles.base, className, {
    [styles.base__skeleton]: !isVisible,
  })

  return (
    <a
      className={baseClasses}
      href={link}
      target="_blank"
      rel="noreferrer"
      ref={ref}
    >
      <img
        loading="lazy"
        src={src}
        alt={alt}
        width={width}
        height={height}
      />
    </a>
  )
})
