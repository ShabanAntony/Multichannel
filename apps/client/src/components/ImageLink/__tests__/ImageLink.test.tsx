import React from 'react'
import renderer from 'react-test-renderer'

import * as utils from '../../../utils/VisibilityObserver'
import { ImageLink } from '../ImageLink'

describe('ImageLink', () => {
  it('should renders correct', () => {
    const tree = renderer
      .create(
        <ImageLink
          link="link"
          src="src"
          alt="alt"
          width={1}
          height={1}
        />,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should renders when visible', () => {
    const spy = jest
      .spyOn(utils, 'useVisibilityObserver')
      .mockReturnValue([{} as any, true])

    const tree = renderer
      .create(
        <ImageLink
          link="link"
          src="src"
          alt="alt"
          width={1}
          height={1}
        />,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(spy).toHaveBeenCalled()
  })
})
