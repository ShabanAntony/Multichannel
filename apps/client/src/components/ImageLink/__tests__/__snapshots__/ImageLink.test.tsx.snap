// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLink should renders correct 1`] = `
<a
  className="base base__skeleton"
  href="link"
  rel="noreferrer"
  target="_blank"
>
  <img
    alt="alt"
    height={1}
    loading="lazy"
    src="src"
    width={1}
  />
</a>
`;

exports[`ImageLink should renders when visible 1`] = `
<a
  className="base"
  href="link"
  rel="noreferrer"
  target="_blank"
>
  <img
    alt="alt"
    height={1}
    loading="lazy"
    src="src"
    width={1}
  />
</a>
`;
